### PR TITLE
nimble/host: Fix potential MITM vulnerability in case of secure connection pairing

### PR DESCRIPTION
**Pull request by prasad-alatkar**
Issue: In case of secure connection pairing and Paskkey Entry method, there is chance that MITM attacker can reply with same public key as received from vulnerable BLE device. This can potentially authenticate MITM attacker. This PR fixes this potential vulnerability.
 
Solution: Check if the peer public key is same as our generated public key. Return fail if the public keys match to avoid 'Impersonation in Passkey entry protocol' vulnerability.

### CODE REVIEW

**Comment by prasad-alatkar** [comment]
@sjanc @rymanluk Could you guys take a look ?

**Comment by sjanc** [comment]
do I get it right that without this patch pairing would end with success even though remote peer is not knowing our private key?
How would it calculate valid DHKey and then mackey and LTK?

File: nimble/host/src/ble_sm_sc.c [hunk 1]
```diff
@@ -612,6 +612,14 @@ ble_sm_sc_public_key_rx(uint16_t conn_handle, struct os_mbuf **om,
     }
 
     cmd = (struct ble_sm_public_key *)(*om)->om_data;
+    /* Check if the peer public key is same as our generated public key.
+     * Return fail if the public keys match. */
+    if (ble_sm_sc_keys_generated) {
+        if (memcmp(cmd, ble_sm_sc_pub_key, 64) == 0) {
+            res->enc_cb = 1;
+            res->sm_err = BLE_SM_ERR_AUTHREQ;
```
Comment (rymanluk) [inline nimble/host/src/ble_sm_sc.c:620]: shouldn't be here a return?

Could you please create a unit test for that in nimble/host/test/src/ble_sm_sc_test.c

**Review by rymanluk — CHANGES_REQUESTED** [review]

File: nimble/host/src/ble_sm_sc.c [hunk 2]
```diff
@@ -612,6 +612,14 @@ ble_sm_sc_public_key_rx(uint16_t conn_handle, struct os_mbuf **om,
     }
 
     cmd = (struct ble_sm_public_key *)(*om)->om_data;
+    /* Check if the peer public key is same as our generated public key.
+     * Return fail if the public keys match. */
+    if (ble_sm_sc_keys_generated) {
```
Comment (sjanc) [inline nimble/host/src/ble_sm_sc.c:617]: I think this check is not needed as ble_sm_sc_ensure_keys_generated() is called before

Comment (prasad-alatkar) [reply nimble/host/src/ble_sm_sc.c:617]: Yes, the check is not needed.

File: nimble/host/src/ble_sm_sc.c (hunk 1 above)
Comment (prasad-alatkar) [reply nimble/host/src/ble_sm_sc.c]: Added return. I will create unit test for the same.

**Review by sjanc — APPROVED** [review]

**Review by rymanluk — COMMENTED** [review]
Looks good. Any chance for the unit test for that ?:)

**Comment by prasad-alatkar** [comment]
> Looks good. Any chance for the unit test for that ?:)

@rymanluk yes, I will add the unit test by this weekend :)

**Comment by prasad-alatkar** [comment]
> > Looks good. Any chance for the unit test for that ?:)
> 
> @rymanluk yes, I will add the unit test by this weekend :)


@rymanluk, I have tried to add unit test to verify the scenario. PTAL and suggest if any improvements can be done...

**Comment by dhrishi** [comment]
Hi @rymanluk Can you please review the unit test and merge if it looks okay? Post that, we will pull this into Espressif's NimBLE repository.

**Review by rymanluk — APPROVED** [review]
Thanks! Looks good to me.

**Pull request merged**